### PR TITLE
Update `GcsBucket` to implement `ObjectStorageBlock` interface

### PR DESCRIPTION
**Pull request by ahuang11**
Implements the new abstract ObjectStorageBlock interface. Will need to figure out how to deprecate things...

Closes (should we create a tracking issue for all the blocks that need refactoring?)

### Screenshots

Do we have a DEPRECATED banner somewhere? Also no idea where the __json_encoder__ is coming from.
image

### Checklist

- [ ] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-gcp/issues/new/choose) first.
- [x] Includes tests or only affects documentation.
- [x] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [x] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-gcp/blob/main/CHANGELOG.md)

### CODE REVIEW

File: prefect_gcp/cloud_storage.py [hunk 1]
```diff
@@ -665,3 +665,219 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+
+    gcp_credentials: GcpCredentials
+    bucket_name: str
+
+    async def get_bucket(self) -> "Bucket":
+        client = self.gcp_credentials.get_cloud_storage_client()
+        bucket = await run_sync_in_worker_thread(client.get_bucket, self.bucket_name)
+        return bucket
+
+    async def list_blobs(self, folder) -> HTTPIterator:
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:684]: Not sure how to type this properly

File: prefect_gcp/cloud_storage.py [hunk 2]
```diff
@@ -665,3 +665,219 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+
+
+class Gcs(ObjectStorageBlock):
+    """
+    Block that represents a resource that can upload and download
+    objects in GCP Cloud Storage Buckets.
+    """
+
+    gcp_credentials: GcpCredentials
+    bucket_name: str
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:677]: Need field.

File: prefect_gcp/cloud_storage.py [hunk 3]
```diff
@@ -665,3 +665,219 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+        to_folder: str,
+        **upload_kwargs: Dict[str, Any],
+    ) -> str:
+        """
+        Uploads a folder to the object storage service from a path.
+
+        Args:
+            from_folder: The path to the folder to upload from.
+            to_folder: The path to upload the folder to.
+            **upload_kwargs: Additional keyword arguments to pass to
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:868]:
```suggestion
            **upload_kwargs: Additional keyword arguments to pass to.
```

File: prefect_gcp/cloud_storage.py [hunk 4]
```diff
@@ -665,3 +665,333 @@ async def write_path(self, path: str, content: bytes) -> str:
                 gcp_credentials=self.gcp_credentials,
             )
         return path
+
+
+class Gcs(ObjectStorageBlock):
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:670]: Probably shouldn't overwrite the core library

File: prefect_gcp/cloud_storage.py [hunk 5]
```diff
@@ -665,3 +665,333 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+        """
+        from_folder = Path(from_folder)
+
+        if to_folder is None:
+            to_folder = Path(from_folder.stem)
+        else:
+            to_folder = Path(to_folder)
+
+        bucket = await self.get_bucket()
+        for from_path in from_folder.glob("**/*"):
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:989]: Is this recursive?

File: prefect_gcp/cloud_storage.py [hunk 6]
```diff
@@ -665,3 +665,333 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+        """
+        client = self.gcp_credentials.get_cloud_storage_client()
+        bucket = await run_sync_in_worker_thread(client.get_bucket, self.bucket_name)
+        return bucket
+
+    # QUESTION: should methods like this return original objects?
+    @sync_compatible
+    async def list_blobs(self, folder: str = "") -> List[Blob]:
+        """
+        Lists all blobs in the bucket that are in a folder.
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:711]: Should mention skip directory.

**Review by desertaxle — COMMENTED** [review]
Overall this looks good! A couple of additional points:
- Do we need to bump the minimum `prefect` version?
- I think the new methods should in corporate use of `bucket_folder` because that seems useful.

File: prefect_gcp/cloud_storage.py [hunk 7]
```diff
@@ -665,3 +665,319 @@ async def write_path(self, path: str, content: bytes) -> str:
                 gcp_credentials=self.gcp_credentials,
             )
         return path
+
+    # NEW INTERFACE METHODS BELOW
+    @sync_compatible
+    async def get_bucket(self) -> "Bucket":
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply prefect_gcp/cloud_storage.py:691]: I imagine eventually, this block will encapsulate all the existing cloud_storage.py tasks.

File: prefect_gcp/cloud_storage.py [hunk 8]
```diff
@@ -665,3 +665,319 @@ async def write_path(self, path: str, content: bytes) -> str:
                 gcp_credentials=self.gcp_credentials,
             )
         return path
+
+    # NEW INTERFACE METHODS BELOW
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply prefect_gcp/cloud_storage.py]: I like it to help developer demarcate (wow I'm surprised I spelled this word correctly)

File: prefect_gcp/cloud_storage.py [hunk 9]
```diff
@@ -665,3 +665,351 @@ async def write_path(self, path: str, content: bytes) -> str:
                 gcp_credentials=self.gcp_credentials,
             )
         return path
+
+    # DEPRECATED MNEW BLOCK INTERFACE METHODS BELOW
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:669]:
```suggestion
    # NEW BLOCK INTERFACE METHODS BELOW
```

**Comment by ahuang11** [comment]
Should we expose GcsBucket at top level init?

File: prefect_gcp/cloud_storage.py [hunk 10]
```diff
@@ -665,3 +665,391 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+        )
+
+        await run_sync_in_worker_thread(
+            blob.upload_from_filename, filename=from_path, **upload_kwargs
+        )
+        return bucket_path
+
+    @sync_compatible
+    async def upload_from_file_object(
+        self, from_file_object: BinaryIO, to_path: str, **upload_kwargs
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:951]: Do we need ignore files?

File: prefect_gcp/cloud_storage.py [hunk 11]
```diff
@@ -665,3 +665,391 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+            to_path = to_folder / bucket_path
+            to_path.parent.mkdir(parents=True, exist_ok=True)
+            self.logger.info(
+                f"Downloading blob from bucket {self.bucket!r} path {bucket_path!r}"
+                f"to {to_path}."
+            )
+            await run_sync_in_worker_thread(
+                blob.download_to_filename, filename=str(to_path), **download_kwargs
+            )
+        return to_folder
```
Comment (ahuang11) [inline prefect_gcp/cloud_storage.py:901]: Should these be absolute paths?

File: prefect_gcp/cloud_storage.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (desertaxle) [reply prefect_gcp/cloud_storage.py]: That's fair. In that case, can you add start and stop comments around just the `ObjectStorageBlock` methods and explicitly call out that they are from the `ObjectStorageBlock` interface?

e.g.
```python
# ----------------------------------------------
# Start ObjectStorageBlock methods
# ----------------------------------------------

...

# ----------------------------------------------
# End ObjectStorageBlock methods
# ----------------------------------------------
```

**Comment by desertaxle** [comment]
> Should we expose GcsBucket at top level init?

I don't see a new reason to expose `GcsBucket` at the top level. But if we do, we should expose all the blocks in the top level to be consistent.

File: prefect_gcp/cloud_storage.py (hunk 10 above)
Comment (desertaxle) [reply prefect_gcp/cloud_storage.py:951]: I don't think so. That was mainly for deployments and these methods aren't going to be use for deployment storage.

File: prefect_gcp/cloud_storage.py [hunk 12]
```diff
@@ -665,3 +665,391 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+                continue
+            to_path = to_folder / bucket_path
+            to_path.parent.mkdir(parents=True, exist_ok=True)
+            self.logger.info(
+                f"Downloading blob from bucket {self.bucket!r} path {bucket_path!r}"
+                f"to {to_path}."
+            )
+            await run_sync_in_worker_thread(
+                blob.download_to_filename, filename=str(to_path), **download_kwargs
+            )
```
Comment (desertaxle) [inline prefect_gcp/cloud_storage.py:900]: Could be worth using `asyncio.gather` to download files concurrently.

File: prefect_gcp/cloud_storage.py (hunk 11 above)
Comment (desertaxle) [reply prefect_gcp/cloud_storage.py:901]: I think that would be more explicit and less prone to error.

File: prefect_gcp/cloud_storage.py [hunk 13]
```diff
@@ -665,3 +665,391 @@ async def write_path(self, path: str, content: bytes) -> str:
...
+                continue
+            bucket_path = str(Path(bucket_folder) / from_path.relative_to(from_folder))
+            self.logger.info(
+                f"Uploading from {str(from_path)!r} to the bucket "
+                f"{self.bucket!r} path {bucket_path!r}."
+            )
+            blob = bucket.blob(bucket_path)
+            await run_sync_in_worker_thread(
+                blob.upload_from_filename, filename=from_path, **upload_kwargs
+            )
```
Comment (desertaxle) [inline prefect_gcp/cloud_storage.py:1050]: Could be worthwhile to use `asyncio.gather` to upload files concurrently.

**Review by desertaxle — APPROVED** [review]

**Review by desertaxle — APPROVED** [review]

**Pull request merged**